gpio: gpio_sx1509b: add support for interrupt functionality

sviraaj opened the pull request:
enable the interrupt functionality for sx1509b gpio
expander, when the CONFIG_GPIO_SX1509B_INTERRUPT config
is enabled. The gpio pin used for interrupt should be
configured in the device tree sx1509b node before enabling
the interrupt configuration.

Signed-off-by: Viraaj Somayajula <sviraaj@zedblox.com>

sviraaj committed:
gpio: gpio_sx1509b: add support for interrupt functionality

enable the interrupt functionality for sx1509b gpio expander,
when the CONFIG_GPIO_SX1509B_INTERRUPT config is enabled.
The gpio pin used for interrupt should be configured in the
device tree sx1509b node before enabling the interrupt
configuration.

Signed-off-by: Viraaj Somayajula <sviraaj@zedblox.com>
Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot commented on dts/bindings/gpio/semtech,sx1509b-gpio.yaml:
```diff
@@ -18,6 +18,13 @@ properties:
       required: true
       const: 16
 
+    int-gpios:
```
Please name this `nint-gpios` to match the datasheet signal name `NINT`.

sviraaj replied:
Done.

pabigot commented on drivers/gpio/gpio_sx1509b.c:
```diff
@@ -348,6 +537,43 @@ static int sx1509b_init(struct device *dev)
+	drv_data->interrupt_sup = 0;
+#endif
+
+	/* Reset state */
```
Remove this and instead add this to the existing initialization after the `k_sleep()` below:

    .interrupt = ALL_PINS,

sviraaj replied:
Can you please help me understand this? By default we are enabling all interrupts?

pabigot replied:
The `interrupt` field corresponds to `RegInterruptMask` which powers up with all ones.  Interrupts must be enabled by clearing the corresponding bit.

I hadn't bothered initializing the field properly because we didn't have interrupt support.  Now it needs to be initialized.

pabigot commented on drivers/gpio/gpio_sx1509b.c:
```diff
@@ -348,6 +537,43 @@ static int sx1509b_init(struct device *dev)
+	drv_data->gpio_int = device_get_binding(cfg->gpio_int_dev_name);
+	if (!drv_data->gpio_int) {
+		/* interrupt not supported */
+		drv_data->interrupt_sup = 0;
```
This branch should be:

    rc = -ENOTSUP;
    goto out;

because if interrupt configuration is enabled the binding has to be present, and if it's present and the pin doesn't exist the configuration is wrong and the driver should fail.

Ideally we wouldn't have `CONFIG_GPIO_SX1509B_INTERRUPT`, and would instead use the existence of the`nint-gpios` devicetree property.

Then you can get rid of the `interrupt_sup` flag (which should have been `bool` not `u8_t`).

sviraaj replied:
I agree. Using nint-gpios property automatically is better. I will try that next. Right now, I have retained CONFIG_GPIO_SX1509B_INTERRUPT. Thanks.

pabigot commented on drivers/gpio/gpio_sx1509b.c:
```diff
@@ -348,6 +537,43 @@ static int sx1509b_init(struct device *dev)
+	} else {
+		drv_data->interrupt_sup = 1;
+
+		drv_data->work.handler = sx1509b_work_handler;
```
Use `k_work_init().`

sviraaj replied:
Done.

pabigot commented on drivers/gpio/gpio_sx1509b.c:
```diff
@@ -320,12 +460,55 @@ static int pin_interrupt_configure(struct device *dev,
 
-	if (mode != GPIO_INT_MODE_DISABLED) {
-		ret = -ENOTSUP;
+	if ((mode != GPIO_INT_MODE_DISABLED) &&
```
This should be:

	if (!IS_ENABLED(CONFIG_GPIO_EX1509B_INTERRUPT)
	    && (mode != GPIO_INT_MODE_DISABLED)) {
		return -ENOTSUP;
	}

sviraaj replied:
Done!

pabigot reviewed (changes requested):
A few preliminary comments.  The interrupt infrastructure is not working for me; when you've addressed these and confirm it works for you I'll revisit it.

BTW: The `NINT` signal is not connected on thingy52 so that target can't use interrupt support.

sviraaj commented:
> 
> 
> A few preliminary comments. The interrupt infrastructure is not working for me; when you've addressed these and confirm it works for you I'll revisit it.
> 
> BTW: The `NINT` signal is not connected on thingy52 so that target can't use interrupt support.

I have verified this with my custom board, after the latest commit. The interrupt is working. Tested with a button and altered the button sample code with the necessary flags to test it. Thanks.

pabigot committed:
tests: drivers: gpio: add SX1509B interrupt-supporting configs

No in-tree boards support the NINT signal on an SX1509B IO extender,
so test using custom overlays/configurations for the Particle Xenon
board using an SX1509B breakout board.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot commented on drivers/gpio/gpio_sx1509b.c:
```diff
@@ -112,6 +148,83 @@ static inline int i2c_reg_write_word_be(struct device *dev, u16_t dev_addr,
+
+	int_source = sys_be16_to_cpu(int_source);
+
+	if ((int_source & drv_data->cb_pins) != 0) {
```
You can't invoke the callbacks with the semaphore held, because the callbacks might (generally will) use the GPIO device to reconfigure the interrupt conditions.

pabigot commented on drivers/gpio/gpio_sx1509b.c:
```diff
@@ -320,12 +459,55 @@ static int pin_interrupt_configure(struct device *dev,
+	k_sem_take(&drv_data->lock, K_FOREVER);
+
+	if (mode == GPIO_INT_MODE_EDGE) {
+		pins->interrupt &= ~BIT(pin);
```
The `gpio_{en,dis}able_callback()` API, which also adjusts `cb_pins`, is obsolete, so as long as invoking the callback depends on a match between detected sense and enabled pins this function must also maintain `cb_pins`.

pabigot reviewed (changes requested):
OK, I've checked this out.  I ran into a variety of problems getting the in-tree test case to work.  Most of the issues are mentioned as review comments, but they are also fixed and more completely described in https://github.com/pabigot/zephyr/commits/rework/24303

At a minimum the two `SQUASH` commits in that branch plus your change removing `interrupt_sup` should be squashed into the single commit in this PR.  Optionally the commit that adds support for testing SX1509B with a breakout board could be added, or I could do that in a separate PR.

Please take a look at that and test it in your environment.  If you wish, I can do the squash and update your branch for you.

sviraaj commented on drivers/gpio/gpio_sx1509b.c:
```diff
@@ -112,6 +148,83 @@ static inline int i2c_reg_write_word_be(struct device *dev, u16_t dev_addr,
+				 SX1509B_REG_INTERRUPT_SOURCE,
+				 (u8_t *)&int_source,
+				 sizeof(int_source));
+	if (ret) {
```
Thanks for pointing this out. Followed the old driver coding style when I first edited this driver, and missed this after moving to latest tree.

sviraaj commented:
> 
> 
> OK, I've checked this out. I ran into a variety of problems getting the in-tree test case to work. Most of the issues are mentioned as review comments, but they are also fixed and more completely described in https://github.com/pabigot/zephyr/commits/rework/24303
> 
> At a minimum the two `SQUASH` commits in that branch plus your change removing `interrupt_sup` should be squashed into the single commit in this PR. Optionally the commit that adds support for testing SX1509B with a breakout board could be added, or I could do that in a separate PR.
> 
> Please take a look at that and test it in your environment. If you wish, I can do the squash and update your branch for you.

@pabigot I checked the rework branch out and tested it using my setup. The interrupts are working as expected for me. You can go ahead with your rework branch and squashing the commits in this PR. Please let me know if there is anything you need from me. Thank you.

pabigot commented:
Unfortunately I do not have permissions required to update this PR as a maintainer.  I've pushed what should be the update to https://github.com/pabigot/zephyr/commits/rework/24303.  If I can get the permissions fixed, I'll update this; or you can update this; or we can close this PR and I'll resubmit from mine.

pabigot commented:
OK, figured out my error and did the push.  Thanks @carlescufi for providing test space.
